### ISSUE TITLE
hTryLock: lock already exists: /home/simon/.cabal/packages/hackage.haskell.org/hackage-security-lock

### ISSUE THREAD

**Issue by sjakobi** [haskell/cabal#6222]
```
$ cabal install cabal2nix -w ghc-8.6.5
Resolving dependencies...
Build profile: -w ghc-8.6.5 -O1
In order, the following will be built (use -v for more details):
 - Glob-0.10.0 (lib) (requires build)
 - attoparsec-0.13.2.3 (lib) (requires build)
 - connection-0.3.1 (lib:connection) (requires download & build)
 - hopenssl-2.2.4 (lib) (requires download & build)
 - language-nix-2.1.0.1 (lib) (requires download & build)
 - tar-0.5.1.1 (lib) (requires download & build)
 - aeson-1.4.4.0 (lib) (requires build)
 - http-client-tls-0.3.5.3 (lib) (requires build)
 - yaml-0.11.1.2 (lib) (requires build)
 - hackage-db-2.0.1 (lib) (requires download & build)
 - distribution-nixpkgs-1.2 (lib) (requires download & build)
 - hpack-0.32.0 (lib) (requires download & build)
 - cabal2nix-2.14.4 (lib) (requires download & build)
 - cabal2nix-2.14.4 (exe:hackage2nix) (requires download & build)
 - cabal2nix-2.14.4 (exe:cabal2nix) (requires download & build)
Downloading  hopenssl-2.2.4
Starting     Glob-0.10.0 (lib)
Starting     attoparsec-0.13.2.3 (lib)
Downloaded   hopenssl-2.2.4
Downloading  tar-0.5.1.1
Starting     hopenssl-2.2.4 (lib)
Building     Glob-0.10.0 (lib)
Building     attoparsec-0.13.2.3 (lib)
Downloaded   tar-0.5.1.1
Downloading  connection-0.3.1
Starting     tar-0.5.1.1 (lib)
Downloading  hackage-db-2.0.1
Building     hopenssl-2.2.4 (lib)
Building     tar-0.5.1.1 (lib)
Downloading  hpack-0.32.0
Installing   hopenssl-2.2.4 (lib)
Installing   Glob-0.10.0 (lib)
Completed    hopenssl-2.2.4 (lib)
Downloading  language-nix-2.1.0.1
Completed    Glob-0.10.0 (lib)
Downloading  distribution-nixpkgs-1.2
Downloading  cabal2nix-2.14.4
Installing   tar-0.5.1.1 (lib)
Completed    tar-0.5.1.1 (lib)
Installing   attoparsec-0.13.2.3 (lib)
Completed    attoparsec-0.13.2.3 (lib)
cabal: Failed to download connection-0.3.1 (which is required by
exe:hackage2nix from cabal2nix-2.14.4 and exe:cabal2nix from
cabal2nix-2.14.4). The exception was:
user error (hTryLock: lock already exists:
/home/simon/.cabal/packages/hackage.haskell.org/hackage-security-lock)
Failed to download language-nix-2.1.0.1 (which is required by exe:hackage2nix
from cabal2nix-2.14.4 and exe:cabal2nix from cabal2nix-2.14.4). The exception
was:
user error (hTryLock: lock already exists:
/home/simon/.cabal/packages/hackage.haskell.org/hackage-security-lock)
```

Re-running the same command then succeeded.

This seems related to https://github.com/haskell/cabal/issues/4960, but in my case there was no second `cabal` process downloading packages, only a long-running `cabal repl`.

```
$ cabal --version
cabal-install version 3.1.0.0
compiled using version 3.1.0.0 of the Cabal library
```

(This is the current `cabal-install-head` from hvr's repo.)

**Comment by 23Skidoo** [comment]
How often does this happen?

**Comment by sjakobi** [comment]
> How often does this happen?

This was the first time for me.

**Comment by 23Skidoo** [comment]
I think that the package downloading code has some parallelism in it, maybe there's a deadlock in it somewhere.

**Comment by bgamari** [comment]
Indeed I am also seeing this in the new [head.hackage CI infrastructure](https://gitlab.haskell.org/ghc/head.hackage/merge_requests/40).

**Comment by mpickering** [comment]
I've also seen this happen on GHC CI.

**Comment by phadej** [comment]
May or may not be fixed when https://github.com/haskell/hackage-security/issues/228 is resolved.

**Comment by bgamari** [comment]
@phadej I doubt this https://github.com/haskell/hackage-security/issues/228 will affect this. Afterall, the locking implementation from `base` is already used when it's available. That issue will only affect the fallback path.

**Comment by bgamari** [comment]
Alright, with the help of @gbaz and @dcoutts I've fixed the likely cause of this in https://github.com/haskell/hackage-security/pull/230.

**Comment by bgamari** [comment]
I believe this should be fixed with the next `hackage-security` release.

**Comment by hvr** [comment]
> I believe this should be fixed with the next `hackage-security` release.

Which will also require a new cabal release... ;-)

**Comment by hanshoglund** [comment]
Also seing this intermittently, e.g. in https://app.circleci.com/jobs/github/hanshoglund/music-suite/94

**Comment by phadej** [comment]
That page is unviewable (at least using mobile browser), it displays fancy 404

**Comment by hanshoglund** [comment]
> That page is unviewable 

Thanks, seems I have a CircleCI issue. Anyway error message is the same as above.

**Comment by phadej** [comment]
should be fixed in 3.0.1.0 / 3.2.0.0 (see `-rc` tags).

**Comment by symbiont-sam-halliday** [comment]
Great work finding the cause! :clap: Is there an ETA for a 3.0.1.0 release?
